draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -1303,78 +1345,65 @@ Steps:
+obtain a point on the twisted Edwards curve.
 
-Preconditions: A Twisted Edwards curve.
+For a twisted Edwards curve given by a * x^2 + y^2 = 1 + d * x^2 * y^2,
+first compute A and B, the parameters of the equivalent Montgomery curve,
+as follows:
 
-Input: alpha, an octet string to be hashed.
+- A = (a + 2) / 2
+- B = (a - d)^2 / 16
 
-Constants:
+Next, use A and B as the curve parameters in the Elligator 2 method of
+{{elligator2}} to obtain a point (x', y') on the Montgomery curve.
+Finally, convert (x', y') to a point (x, y) on the target curve.
+Letting B' = 4 / (a - d), compute
 
-- A and D, the parameters of the curve
+- x = x' / y'
+- y = (B' * x' - 1) / (B' * x' + 1)
 
-- Z, the smallest (in absolute value) non-square in F, breaking ties by choosing
-  the positive value.
+This can be done in one inversion using Montgomery's trick {{M87}}:
+invert the product y' * (B' * x' + 1), then multiply by y' to obtain
+1 / (B' * x' + 1), and likewise for 1 / y'.
 
-Output: (x, y), a point on E.
+Preconditions: A twisted Edwards curve.
 
-Operations:
+Input: alpha, an octet string to be hashed.
 
-~~~
-1.  u = hash2base(alpha)
-2. c1 = 2 * (A + D) / (A - D)
-3. t1 = -c1 / (1 + Z * u^2)
-4. t2 = -t1 - c1
-5. g1 = t1^3 + c1 * t1^2 + t1
-6. If is_square(g1), set t = t1, else t = t2
-7. y = (t - 1) / (t + 1)
-8. x = sqrt((y^2 - 1) / (D * y^2 - A), q)
-9. Output h * (x, y)
-~~~
+Constants:
 
-#### Implementation
+- A and B, the parameters of the equivalent Montgomery curve, and B' = 1 / sqrt(B).
 
-The following procedure implements elligator2 method in a straight-line
-fashion.
+- Z, the smallest (in absolute value) non-square in F, breaking ties by choosing
+  the positive value.
 
-~~~
-map2curve_ell2edwards(alpha)
-Input: alpha, an octet string to be hashed.
 Output: (x, y), a point on E.
 
-Constants:
-1. c0 is an non-square in F.
-2. c1 = 2 * (A + D) / (A - D).
-
-Steps:
-1.   u = hash2base(alpha)
-2.  t1 = u^2
-3.  t1 = c0 * t1
-4.  t1 = t1 + 1
-5.  t1 = inv0(t1)
-6.  t1 = t1 * c1
-7.  t1 = -t1                // t1 = -c1 / (1 + c0 * u^2)
-8.  t2 = -t1 - c1           // t2 = -x1 - c1
-9.  g1 = t1 + c1
-10. g1 = g1 * t1
-11. g1 = g1 + 1
-12. g1 = g1 * t1            // g1 = t1^3 + c1 * t1^2 + t1
-13.  e = is_square(g1, q)
-14. t3 = CMOV(t2, t1, e)    // If e=True, t=t1, else t=t2
-15. t4 = t3 + 1
-16. t4 = inv0(t4)
-17.  y = t3 - 1
-18.  y = y * t4
-19. t5 = y^2
-20. gx = t5 * D
-21. gx = gx - A
-22. gx = inv0(gx)
-23. t5 = t5 - 1
-24. gx = gx * t5
-25.  x = sqrt(gx, q)
-26. Output h * (x, y)
+Exceptions: The exceptions for the Elligator 2 encoding are as given in
+{{elligator2}}. When converting to a point on the twisted Edwards curve, the remaining exceptions
+are y' == 0 or B' * x' == -1. Implementors must detect these cases and return (x, y) = (0, 1).
+
+The following straight-line implementation handles the exceptional cases:
+
+~~~
+1. (x', y') = map2curve_elligator2(alpha)  // a Montgomery point
```

Comment:
Should we generalize the Montgomery->Edwards point conversion function? (I would have benefited from such a function being written in cleartext *somewhere* earlier this year.) We might then also include conversion from, say, Weierstrass to Montgomery, among others.

Reply:
Opened #107 about this

Note that while Montgomery -> Weierstrass is easy, the opposite isn't guaranteed to exist, but it's relatively easy to write down the necessary and sufficient conditions for its existence (due to [Okeya, Kurumatani, and Sakurai, PKC 2000](https://doi.org/10.1007%2F978-3-540-46588-1_17))